pkg/monitor/ovsdb.go

```diff
@@ -160,411 +162,424 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
-	if !ok {
-		return nil
+	if ok && ofPort > 0 {
+		monitor.endpointMap[uuid].PortNo = uint32(ofPort)
 	}
 
-	if ofPort <= 0 {
-		// Parsing added ofport error: invalid local endpoint ofPort. In OfPort initializing status
-		return nil
+	_, err := net.ParseMAC(macStr)
+	if err != nil {
+		macStr = ""
 	}
-	ofport := uint32(ofPort)
+	monitor.endpointMap[uuid].MacAddrStr = macStr
 
-	macAddr, err := net.ParseMAC(macStr)
-	if err != nil {
-		klog.Errorf("Parsing endpoint macAddr error: %v", macAddr)
-		return nil
+	// if endpoint info is ready, trigger endpoint add callback
+	if monitor.isEndpointReady(monitor.endpointMap[uuid]) {
+		monitor.ovsdbEventHandler.AddLocalEndpoint(monitor.endpointMap[uuid])
 	}
-	monitor.localEndpointHardwareAddrCache[macStr] = uint32(ofPort)
+}
 
-	if err := monitor.waitForPortPresent(rowupdate.New.Fields["name"].(string), 1*time.Second); err != nil {
-		klog.Errorf("Failed to add local endpoint, wait ovs port present timeout, error: %v", err)
-		return nil
+func (monitor *OVSDBMonitor) processOvsPortUpdate(uuid string, rowupdate ovsdb.RowUpdate) {
+	var newEndpoint, oldEndpoint *datapath.Endpoint
+	var newIfaceUUID, oldIfaceUUID string
+
+	oldIfaces := listUUID(rowupdate.Old.Fields["interfaces"])
+	newIfaces := listUUID(rowupdate.New.Fields["interfaces"])
+	if len(newIfaces) > 1 || len(oldIfaces) > 1 {
+		// bond port
+		return
 	}
 
-	return monitor.interfaceToEndpoint(ofport, rowupdate.New.Fields["name"].(string), macStr)
-}
+	newIfaceUUID = newIfaces[0].GoUuid
+	if len(oldIfaces) == 0 {
+		oldIfaceUUID = newIfaceUUID
+	} else {
+		oldIfaceUUID = oldIfaces[0].GoUuid
+	}
 
-func (monitor *OVSDBMonitor) waitForPortPresent(interfaceName string, timeout time.Duration) error {
-	return wait.PollImmediate(10*time.Millisecond, timeout, func() (do bool, err error) {
-		if !monitor.isPortExists(interfaceName) {
-			return false, nil
+	if oldIfaceUUID != newIfaceUUID {
+		// ovsport interfaces field update
+		if _, ok := monitor.endpointMap[oldIfaceUUID]; !ok {
+			monitor.endpointMap[oldIfaceUUID] = &datapath.Endpoint{}
+		} else {
+			oldEndpoint = monitor.endpointMap[oldIfaceUUID]
 		}
 
-		return true, nil
-	})
-}
+		if _, ok := monitor.endpointMap[newIfaceUUID]; !ok {
+			monitor.endpointMap[newIfaceUUID] = &datapath.Endpoint{}
+		} else {
+			newEndpoint = monitor.endpointMap[newIfaceUUID]
+		}
 
-func (monitor *OVSDBMonitor) isPortExists(interfaceName string) bool {
-	monitor.cacheLock.RLock()
-	defer monitor.cacheLock.RUnlock()
-	for _, port := range monitor.ovsdbCache["Port"] {
-		if port.Fields["name"].(string) == interfaceName {
-			return true
+		// Is this case exsit
+		if monitor.isEndpointReady(oldEndpoint) && monitor.isEndpointReady(newEndpoint) {
+			monitor.ovsdbEventHandler.UpdateLocalEndpoint(newEndpoint, oldEndpoint)
+		}
+		if monitor.isEndpointReady(newEndpoint) && !monitor.isEndpointReady(oldEndpoint) {
+			monitor.ovsdbEventHandler.AddLocalEndpoint(newEndpoint)
```

Comment:
In L415 ~ 419 if newEndpoint not assign a value, the default value will be "nil", will check the newEndpint cause error?

Reply:
yes, we should init oldendpoint in L410, and init newEndpoint in L416,    I forget it.

actually， this is an exception case which only happens when manually operating OVSDB, it won't normally update the interface uuid in the same port.